ToolChain/Assembler/Parser/Parser.py

```diff
@@ -20,586 +19,393 @@
+            text = text.split(maxsplit=1)
+            instruction += text[1][:-1].encode("utf-8")
+            self.relativeAddressCounter += len(instruction) + 1
+            instruction += b'\x00'
+
+        elif dataIdentifier == DATA_NUMERIC_INDICATOR:
+            # DataNumeric number which must be converted to binary
+            numeric = self.translateTextImmediate(line[1])
+            instruction += struct.pack(">I", numeric)
+            self.relativeAddressCounter += 4
+
+        elif dataIdentifier == DATA_MEMORY_REFERENCE:
+            # Memory reference, label will be returned as the instruction
+            if line[1][0] == MEMORY_REFERENCE_INDICATORS:
+                memref = line[1][1:].upper()
+            else:
+                memref = line[1].upper()
+            instruction += b':' + memref.encode("utf-8") + b':'
+            self.relativeAddressCounter += 4
 
-            # This is so we are able to use an and followed by a right shift to build the instruction
-            # easiest way that my brain could come up with. All other ways of building instructions
-            # seems to rely on ugly logic based on if chaining... Ugly. This is better. We are shifting
-            # until we find the "edge" of this part of the instruction.
+        else:
+            # Identifier not in accepted list
+            raise ValueError("Invalid instruction format at line " + str(self.lineno) + " \n" + dataIdentifier)
 
-            while instructionPartBitMask % 2 is not 1:
-                instructionPartBitMask >>= 1  # Shift right one bit
-                shiftValue += 1
+        return instruction, labelFlag
 
-            # We grab the value of the instruction to determine whether it's an actual instruction or a memory reference
-            instructionPartOriginal = getattr(instruction, instructionPart)
-            # We need an intermediate step because a part could be text
+    def _evaluateFormBasedOnArguments(self, line, form, operandList):
+        """
+        Method looks at the whole line after the initial instruction and determines its form based on the arguments.
+        Registers are concatenated as "REG", immediates and lables as "IMM", etc. We also populate a list of arguments
+        as they are to be constructed into binary code later.
+        :param line: str list, Line split into individual arguments
+        :param form: str, will be built up based on arguments on the line in this method
+        :param operandList: str list, contains our list of arguments to be used later when assembling our binary code
+        :return: str, Fully constructed form and newly populated list of arguments
+        """
 
-            if type(instructionPartOriginal) is str:
-                # This is a memory reference
-                memoryRefSymbol = instructionPartOriginal
-                instructionPartToBeUsed = 0xDEADBEEF
+        opcount = 0
+
+        for arg in line[1:]:
```

Comment:
Switch arg to operand? Not sure. Maybe use a random term like element of whatever.